repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -55,518 +54,373 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
-          request.addHeader(OpenStackServiceHeader.IMPERSONATOR_ROLES, impersonatorRoles.mkString(","))
-        }
-
-        // Construct and add the tenant id header
-        val tenantsToPass = buildTenantHeader(token.defaultTenantId, token.tenantIds, matchedUriTenant)
-        if (tenantsToPass.nonEmpty) {
-          request.addHeader(OpenStackServiceHeader.TENANT_ID, tenantsToPass.mkString(","))
-        }
-
-        // If configured, add roles header
-        if (config.getIdentityService.isSetRolesInHeader && token.roles.nonEmpty) {
-          request.addHeader(OpenStackServiceHeader.ROLES, token.roles.map(_.name).mkString(","))
-        }
-
-        // If present, add the tenant from the URI as part of the Proxy header, otherwise use the default tenant id
-        matchedUriTenant match {
-          case Some(uriTenant) =>
-            request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION, s"$XAuthProxy $uriTenant")
-          case None =>
-            token.defaultTenantId match {
-              case Some(tenant) =>
-                request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION, s"$XAuthProxy $tenant")
-              case None =>
-                request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION, XAuthProxy)
-            }
-        }
-
-        // Construct and add authenticatedBy, if available
-        token.authenticatedBy foreach {
-          _ foreach {
-            request.addHeader(OpenStackServiceHeader.AUTHENTICATED_BY, _)
-          }
-        }
+    def addIdentityStatusHeader(confirmed: Boolean): Unit = {
+      if (Option(config.getDelegating).isDefined) {
+        if (confirmed) request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.CONFIRMED)
+        else request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.INDETERMINATE)
       }
+    }
 
-      def addCatalogHeader(maybeEndpoints: => Try[EndpointsData]): Try[Unit.type] = {
-        if (config.getIdentityService.isSetCatalogInHeader) {
-          maybeEndpoints map { endpoints =>
-            request.addHeader(PowerApiHeader.X_CATALOG, Base64.encodeBase64String(endpoints.json.getBytes))
-            Unit
-          }
-        } else {
-          Success(Unit)
+    /**
+      * BEGIN PROCESSING
+      */
+    val token = getAuthToken
+    lazy val endpoints = token.flatMap(authToken => getEndpoints(authToken)) // Prevents making call if its not needed
+    token flatMap { authToken =>
+      validateToken(authToken) flatMap { validToken =>
+        val authResult = KeystoneV2Authorization.doAuthorization(config, request, validToken, endpoints)
+
+        addTokenHeaders(authResult.scopedToken, authResult.matchedTenant)
+        authResult match {
+          case AuthorizationPassed(scopedToken, _) =>
+            addCatalogHeader(endpoints)
+            addGroupsHeader(getGroups(token.get, scopedToken))
+          case AuthorizationFailed(_, _, exception) => Failure(exception)
```

Comment:
I think you can nest the endpoints a bit more, simplifying this block a small amount:

```
    getAuthToken flatMap { authToken =>
      validateToken(authToken) flatMap { validToken =>
        lazy val endpoints = getEndpoints(authToken) // Prevents making call if its not needed
        val authResult = KeystoneV2Authorization.doAuthorization(config, request, validToken, endpoints)

        addTokenHeaders(authResult.scopedToken, authResult.matchedTenant)
        authResult match {
          case AuthorizationPassed(scopedToken, _) =>
            addCatalogHeader(endpoints)
            addGroupsHeader(getGroups(token.get, scopedToken))
          case AuthorizationFailed(_, _, exception) => Failure(exception)
        }
      }
    }
```

Reply:
Yeah, that's better. I think the current state was from a different set of changes that required it, but as things moved around more it could collapse again.